at_client/lib/src/service/notification_service_impl.dart

```diff
@@ -0,0 +1,236 @@
+import 'package:at_client/src/responseParser/notification_response_parser.dart';
+
+class NotificationServiceImpl implements NotificationService {
+  Map<String, Function> listeners = {};
+  Map<String, StreamController> streamListeners = {};
+  final EMPTY_REGEX = '';
+  static const notificationIdKey = '_latestNotificationId';
+
+  final _logger = AtSignLogger('NotificationServiceImpl');
+
+  late AtClient atClient;
+
+  bool isMonitorStarted = false;
+  Monitor? _monitor;
+  ConnectivityListener? _connectivityListener;
+  NotificationServiceImpl(AtClient atClient) {
+    this.atClient = atClient;
+  }
+
+  void _init() {
+    if (_connectivityListener == null) {
+      _connectivityListener = ConnectivityListener();
+      _connectivityListener!.subscribe().listen((isConnected) {
+        if (isConnected) {
+          _logger.finer(
+              'starting monitor for atsign: ${atClient.getCurrentAtSign()}');
+          _startMonitor();
+        } else {
+          _logger.finer('lost network connectivity');
+        }
+      });
+    }
+  }
+
+  Future<void> _startMonitor() async {
+    if (isMonitorStarted) {
+      _logger.finer('monitor is already started');
+      return;
+    }
+    final lastNotificationTime = await _getLastNotificationTime();
+    _monitor = Monitor(
+        _internalNotificationCallback,
+        _onMonitorError,
+        atClient.getCurrentAtSign()!,
+        atClient.getPreferences()!,
+        MonitorPreference()..keepAlive = true,
+        _monitorRetry);
+    _logger.finer(
+        'starting monitor with last notification time: $lastNotificationTime');
+    await _monitor!.start(lastNotificationTime: lastNotificationTime);
+    isMonitorStarted = true;
+  }
+
+  Future<int?> _getLastNotificationTime() async {
+    final atValue = await atClient.get(AtKey()..key = notificationIdKey);
+    if (atValue.value != null) {
+      _logger.finer('json from hive: ${atValue.value}');
+      return jsonDecode(atValue.value)['epochMillis'];
+    }
+    return null;
+  }
+
+  void stop() {
+    _monitor?.stop();
+    _connectivityListener?.unSubscribe();
+  }
+
+  void _internalNotificationCallback(String notificationJSON) async {
+    // #TODO move some of this logic to notification parser
+    try {
+      var notifications = notificationJSON.split('notification: ');
+      notifications.forEach((notification) async {
+        if (notification.isEmpty) {
+          _logger.finer('empty string in notification');
+          return;
+        }
+        notification = notification.replaceFirst('notification:', '');
+        notification = notification.trim();
+        final atNotification =
+            AtNotification.fromJson(jsonDecode(notification));
+        await atClient.put(AtKey()..key = notificationIdKey, notification);
+        streamListeners.forEach((regex, streamController) {
+          if (regex != EMPTY_REGEX) {
+            final isMatches = regex.allMatches(atNotification.key).isNotEmpty;
```

Comment:
Let's try to avoid temp variables wherever we can.

Reply:
if (regex.allMatches(atNotification.key).isNotEmpty) {
              streamController.add(atNotification);
}

Reply:
done